### ISSUE TITLE
Feature request: Allow Configurable Quote Styling for f-Strings in Python 3.12+

### ISSUE THREAD

**Issue by matthewlloyd** [astral-sh/ruff#14118]
I would like to request an option in Ruff formatter to configure quote styling for f-strings in Python 3.12 and later - specifically, the ability to maintain consistent use of double quotes throughout f-strings, even when nesting quotes.

## Background

With the introduction of Python 3.12, it's now possible to nest double quotes within double-quoted f-strings without needing to alternate between single and double quotes. This enhancement allows for more consistent and uniform code styling.

Currently, Ruff formatter's preview mode reverts nested double quotes to single quotes inside double-quoted f-strings, following the style used in earlier Python versions. While this approach ensures backward compatibility, it may not be ideal for codebases that have migrated to Python 3.12 and prefer consistent quote usage.

## Benefits

* Reduced cognitive load: Consistent quotes within f-strings simplify reading and writing code, reducing the mental overhead associated with switching between quote types.
* Improved readability: Uniform quote usage enhances the clarity of f-strings, especially in longer or more intricate strings with nested quotes.
* Seamless copying and pasting: Expressions within f-strings can be copied and pasted without needing to adjust quote styles, streamlining code modifications.

## Proposal

Introduce a configuration option in Ruff formatter that allows developers to:

* Opt-in to consistent quote styling for f-strings in Python 3.12 and later.
* Choose their preferred quote style (e.g., always use double quotes).

This flexibility would enable developers to take full advantage of the new string handling capabilities in Python 3.12 while adhering to their codebase's stylistic preferences.

This configuration option could be a sibling of the existing option, `quote-style`, which allows developers to choose between double or single quote characters: https://docs.astral.sh/ruff/settings/#format_quote-style

## Alternative Solution

As an alternative to adding a new configuration option, Ruff formatter could automatically switch to the new f-string quote formatting when the target Python version is set to 3.12 or higher. This would involve:

* Default Behavior Change for Python 3.12+: When the target Python version is 3.12 or above, Ruff formatter would default to using consistent quotes throughout f-strings, leveraging the enhanced syntax support.
* Backward Compatibility: For projects targeting earlier Python versions, the formatter would maintain the existing behavior of alternating quotes to ensure compatibility.
* Simplified Configuration: This approach eliminates the need for additional configuration options, making it easier for developers to adopt the new formatting style without extra setup.

## Rationale

* Alignment with Python Evolution: Adopting the new formatting by default for Python 3.12+ aligns Ruff formatter with the latest Python language features and future direction.
* Temporary Compatibility Concerns: While changing quotes might affect syntax highlighting in some tools, major IDEs like VS Code and PyCharm already support the new syntax. Any remaining tooling issues are likely to be resolved over time.
* Future-Proofing: Considering that pre-3.12 versions of Python will reach end-of-life after October 2027, updating the default behavior now helps future-proof codebases and reduces the need for later adjustments.

## Additional context

I understand that changing the default behavior may have implications for existing workflows. However, making this adjustment based on the target Python version provides a balanced approach that benefits developers who are ready to adopt Python 3.12 features while preserving the experience for those who are not.

Thank you for considering this feature request.

## References

For further discussion, see the comments here: https://github.com/astral-sh/ruff/issues/11056#issuecomment-2458269566

The pull request which changed the f-string quote style from consistent to alternating in the formatter's preview mode is here: https://github.com/astral-sh/ruff/pull/13860

**Comment by dhruvmanila** [comment]
(Cross linking for visibility: https://github.com/astral-sh/ruff/discussions/15836)

**Comment by dhruvmanila** [comment]
> pylint is warning about inconsistent quotes, I have to run it with --disable=W1405.

Can you provide an example of where it's triggering this rule? From what I understand reading the documentation of [W1405](https://pylint.pycqa.org/en/latest/user_guide/messages/warning/inconsistent-quotes.html) is that it's more about maintaining the consistency at a global level in the file and not specifically related to f-string but I might be mistaken.

**Comment by vivodi** [comment]
> Pylint's W1405 does check quotes are consistent inside nested f-string expressions when targeting Python 3.12+, see e.g. [pylint-dev/pylint#9113](https://github.com/pylint-dev/pylint/issues/9113).
> 
> Example input (copied from that issue):
> 
> ```
> dictionary = {'0': 0}
> f_string = f'{dictionary["0"]}'
> ```
> 
> Pylint output:
> 
> ```
> ************* Module a
> a.py:2:0: W1405: Quote delimiter " is inconsistent with the rest of the file (inconsistent-quotes)
> ```

I think this justify the PR #16385 @MichaReiser

**Comment by MichaReiser** [comment]
I reply here to some of the comments made in https://github.com/astral-sh/ruff/pull/16385

As stated before, I do think that this option could be useful, but I first want to see more demand because each option has a cost: It requires that every Ruff user makes a decision on what value to pick for this new option. There are options (like quote style) where I think this is justified and others that are useful for many, but still don't meet the bar. That's why I want to hear from more users (by upvoting or explaining their reason for using same quotes if it wasn't discussed before) about why this option meets our high bar for options. 

> I think the Python official decision to introduce 'Quote reuse in f-strings' in [Python 3.12](https://docs.python.org/3/whatsnew/3.12.html#pep-701-syntactic-formalization-of-f-strings) justifies this PR.

I haven't followed the f-string PEP, so I could be wrong here but I don't think that allowing the same quotes suggests in any way that this should now be considered the idiomatic way of quoting f-strings. I see the main motivation as a) allow arbitrary nesting of f-strings and b) be less surprising as the main motivation to allow either quotes in nested f-strings. 

It's also in the very nature of an opinionated formatter to have opinions on what syntax to restrict. For example, Python supports line continuations (lines ending with `\`) but Ruff intentionally removes them in favor of normal line breaks and parentheses. There's no configuration option that allows changing this behavior. This isn't a reason not to add this option, but the opposite isn't true either: Just because Python supports a specific syntax doesn't mean Ruff should have an option for it. 

> If you're worried about providing too many options, why not make this PR (f-string consistent quotes) the only choice for users of Python 3.12 and above?

IMO, It has a poor user experience. Users upgrading their target-version from 3.11 to 3.12 would see a lot of formatting churn that's simply unnecessary. 

> Conclusion: Change the existing behavior to f-string consistent quotes instead of adding a new option. I think this is a reasonable solution, avoiding the issue raised by @MichaReiser about providing too many options.

I don't consider changing the default an option. It creates a lot of churn for users, it has poor backwards compatibility, and IMO, is less readable.

**Comment by vivodi** [comment]
I'll argue for the necessity of *quote reuse in f-strings* from another perspective.  

If a user sets `quote-style = "single"`, it indicates a clear preference for single quotes. In that case, Ruff should strive to use single quotes wherever possible, rather than unnecessarily introducing double quotes in f-strings against the user's preference.  

For example, I find single quotes visually comfortable, but I don't like the look of double quotes.

**Comment by vivodi** [comment]
> It requires that every Ruff user make a decision on what value to pick for this new option.  

Ruff being an opinionated formatter does not prevent it from offering options to accommodate some users' preferences. The "opinionated" aspect of Ruff refers to its default configuration, rather than being a dictatorship based on the personal preferences of Ruff developers. Users who do not care about these formatting details can simply use Ruff’s officially recommended defaults. Meanwhile, for those who dislike the default settings provided by Ruff's developers, Ruff should also cater to their needs.  

Ruff can certainly have opinionated default configurations to ensure a smooth out-of-the-box experience. **However, being opinionated should also come with inclusivity and diversity—it should not turn formatting preferences into a dictatorship of the developers’ own tastes.**

I believe adding an f-string consistent quotes formatting option would not increase maintenance overhead. Users who are not concerned with minor formatting details can stick with the default configuration, while those who dislike inconsistent quote styles should have an option available in Ruff.

**Comment by vivodi** [comment]
> The main reason I hear so far of why such an option might be needed other than just "I prefer it the other way" (which is a fine enough reason) is for compatibility with other tools.

Inconsistent quotes within f-strings violate [Pylint's **W1405**](https://github.com/pylint-dev/pylint/issues/9113) rule. If a user relies on both Pylint and Ruff's formatter, they will encounter Pylint errors unless they explicitly disable W1405.  

Given that Pylint is a **prominent** Python linter with a large user base—and that Ruff's linter has already adopted many of its rules—I believe Ruff's formatter should offer an option that respects Pylint's guidelines. Moreover, the existence of W1405 itself reflects Pylint's stance on this issue: f-strings should maintain **consistent quotes**. This further justifies the necessity of a corresponding option in Ruff.

**Comment by MichaReiser** [comment]
Thanks @vivodi for all your input. I understand your standpoint and I can see that this is very important to you. Let's leave room for other users to express their opinion.

**Comment by trooz** [comment]
I need an option to enforce consistent f-string quotes. This would make batch find-and-replace operations for expressions like `entry['filepath']` simpler and less error-prone.  

Recently, while refactoring my project, I needed to change all `entry['filepath']` values from `str` to `pathlib.Path`. I updated all occurrences of `entry['filepath']` in the project, but later encountered a bug because instances like `f'{entry["filepath"]}'` were not modified. If Ruff enforced consistent f-string quotes, this bug could have been avoided.  

I have already set `quote-style = 'single'`, but `entry["filepath"]` still appears, which eventually caused the bug. I believe this is unreasonable.

**Comment by vasily-v-ryabov** [comment]
Let me add my five cents. We do use `quote-style = "single"` and it works pretty well. But all multi-line strings (which are not a docstrings) are converted to triple double quotes `"""` instead of expected triple single quotes `'''`.

 1. AFAIK a docstring can be easily distinguished from other multi-line string in AST. If one more config option is unwanted, at least non-doc multi-line strings could included into action for existing `quote-style` option. It seems pretty critical to us. We have 1k+ such issues.

 2. Ideal solution would be an option `triple-quote-style`. If it was implemented, the previous case would be much less critical or even unnecessary. So I would support this option for sure. Because the first point is a kind of compromise.

 3. Regarding the way how single quoted strings are re-formatted... Replacing `'\'key\' = \'value\''` with `"'key' = 'value'"` looks obvious, but there are some cases that break quotes consistency like this:
    ```python
                '[ComponentName]: Unknown errors were found for keys:\n'
                "'key': {value!r}\n"
                '{key2!r}: {value2!r}'
    ```
    or for some collections. But such cases are pretty rare (less than 20 in our codebase), they can be handled by `# fmt: off`.

I think the number of necessary `# fmt: off` statements may help everybody to decide how critical their cases or not.

**Comment by jbdyn** [comment]
I am also upvoting for consistent quotation marks in f-strings for the same case @anatoli26 already [mentioned above](https://github.com/astral-sh/ruff/issues/14118#issuecomment-2686702456):

```python
f"Some text {some_dict["something"]}"
```

As @vivodi [wrote before](https://github.com/astral-sh/ruff/issues/14118#issuecomment-2694434187), I like the idea to have this behavior reflected by the `quote-style` setting:

> If a user sets `quote-style = "single"`, it indicates a clear preference for single quotes.

IMO, there is conceptually no difference between inserts in a `.format`, `%` or f-string statement and [PEP 701](https://peps.python.org/pep-0701/) enables one to express this in code:

```python
d = dict(key="foo")

# either
s = "%s or bar" % d["key"]

# or
s = "{} or bar".format(d["key"])

# or
s = f"{d["key"]} or bar"
```

Thanky you all and @matthewlloyd and @MichaReiser for your commitment. Looking forward to see this implemented.

**Comment by lukas-lang** [comment]
Another argument in favor of preserving quotes: Ruff is marketed as a drop-in replacement for Black. However, Black leaves the quotes alone, so switching over to Ruff leads no unexpected formatting changes, forcing you to re-format everything and having a unnecessary "formatted code" commit

**Comment by GideonBear** [comment]
(I initially thought this could sidestep the formatter and be a purely-linter (Q000) issue, but it cannot, because that would cause conflicts with the formatter.)

I would like to propose an alternative solution (that, AFAICT, hasn't been suggested yet):
- Change the formatter to **ignore this altogether**, meaning, leave quotes inside f-string expressions alone completely.
	- This is technically backwards compatible, although I would understand if this is not desired; an alternative would be to make this configurable instead, like `format.keep-quotes-inside-fstrings`.
- Add a new lint rule (Q100 or something) for inconsistent quote usage:
```py
foo = f"{"bar"}"
foo = f"{'bar'}"  # Q100
foo = f'{'bar'}'
foo = f'{"bar"}'  # Q000
```

Then maybe the formatter option could be automatically enabled when that lint rule (Q100) is selected? As the formatter and linter are designed to be used together, I think making them explicitly work together like that is sensible.

Some more thoughts:
- Given the high bar for formatter settings, does it at all help to make a setting disable an opinionated formatting choice, instead of making the setting choose between two options?
- The only reason I haven't made a seperate issue for the (lint) rule suggestion, is because it would conflict with the formatter, and thus be either unusable with the formatter, or require a formatter change anyway. The strict and unconfigurable formatter is holding me back from solving this in my own way (e.g. with pylint, or manually).

I am also +1 for just making this a formatter option, but I wanted to do something with my thought of this being a natural extension of Q000. Maybe this is useful.

P.S., if there is existing discussion or documentation about formatter-linter interaction, I would be interested in reading that

**Comment by MichaReiser** [comment]
I'd prefer this to be a formatter option over disabling formatting entirely. We're also reaching the number of upvotes that I think justify adding the option

**Comment by matthewlloyd** [comment]
I'd like to get input from the folks who 👍'd this issue (and anyone else using the formatter).

There are currently two possible directions for handling quotes in nested strings. I'd like to gauge which one people would actually prefer in practice.

Today, Ruff's formatter optimizes quotes for **escape minimization**, foregoing the user's preferred quote style for the sake of minimizing escapes:

```python
# quote-style = "double"
"a normal string"
"a string with more 'single' quotes than double \""
'a string with more "double" quotes than single \''  # escape minimization
f"an f-string with a {'nested string'}"  # alternation for nested strings
f'an f-string with "double" quotes and a {"nested string"}'  # escape minimization and alternation for nested strings
```

## Option A: use preferred quotes for nested strings in f/t-strings

`nested-string-quote-style = "preferred"`

Add an option that only affects **nested strings inside f/t-strings**, ensuring they always use the preferred quote style. This is the original proposal from this issue and partially implemented in https://github.com/astral-sh/ruff/pull/23098:

```python
# quote-style = "double"
# nested-string-quote-style = "preferred"
"a normal string"
"a string with more 'single' quotes than double \""
'a string with more "double" quotes than single \''  # escape minimization
f"an f-string with a {"nested string"}"  # use preferred quotes for nested strings
f'an f-string with "double" quotes and a {"nested string"}'  # still minimizes escapes for the outer string
```

This keeps Ruff’s current escape-minimization behavior, but makes nested strings in f/t-strings use the preferred quotes.

## Option B: strictly preferred quotes everywhere

`quote-style-strict = true`

Add a more general option that **uses the preferred quote style for all strings**, including nested ones, even if it means more escapes. The thought here is that if you want to see your preferred quotes used for nested strings, you probably want to see them used everywhere else too:

```python
# quote-style = "double"
# quote-style-strict = true
"a normal string"
"a string with more 'single' quotes than double \""
"a string with more \"double\" quotes than single '"  # always use preferred quotes
f"an f-string with a {"nested string"}"  # use preferred quotes for nested strings
f"an f-string with \"double\" quotes and a {"nested string"}"  # always use preferred quotes
```

Rationales:
* Makes the use and meaning of `"`, `'`, `\"` and `\'` consistent across the codebase. Easier to search for specific strings without needing a regex.
* Prevents noisy diffs caused by escape-optimizing quote flips when editing strings.
* Simpler to implement, test, document, and reason about.

Notably, **Option B would subsume Option A**, since nested strings would already follow the preferred quote style; we don't want to add both options.

## Feedback

In lieu of a poll, if you have a preference, please react to this comment:

* 👍 Prefer **Option A** (use preferred quotes for nested strings in f/t-strings)
* 🚀 Prefer **Option B** (strictly preferred quotes everywhere)
* 👀 No strong preference

If you have edge cases, examples from real code, or thoughts on naming the new option, feel free to comment.

**Comment by vivodi** [comment]
Hi @matthewlloyd, thanks for laying out the options so clearly!

I think it would be valuable to include the "holistic consistency" approach mentioned by @MichaReiser earlier in [this comment](https://github.com/astral-sh/ruff/pull/23098#issuecomment-4004764673) as a third option (**Option C**). This would give the community a complete picture of the design space before we settle on a direction.

Here is a draft of how **Option C** could be presented alongside your current ones:

---

## Option C: holistic consistent quotes
`nested-string-quote-style = "consistent"`

Ensure that nested strings **always use the same quote style as the outermost string**. To keep visual noise to a minimum, the formatter evaluates the entire f-string tree (the outer string plus all nested expressions) and chooses the single quote style that results in the fewest total escapes across the whole structure:

```python
# quote-style = "double"
# nested-string-quote-style = "consistent"

"a normal string"
'a string with more "double" quotes than single \''  # normal escape minimization

f"an f-string with a {"nested f-string"}"  # matches outer quotes
f'an f-string with "double" quotes and a {'nested string'}'  # outer flips to single to minimize total escapes for the whole tree
```

This prioritizes a uniform look within f/t-strings, ensuring the quotes remain consistent from the outside in while still respecting Ruff's core philosophy of escape minimization.

**Rationales:**

* **Visual Uniformity:** Guarantees that an f-string and its interpolations use a single, uniform quote style, which is often cited as the primary aesthetic benefit of PEP 701.
* **Smart Minimization:** Unlike Option B, it still optimizes for the fewest backslashes, but it does so holistically. If using your "preferred" quotes would cause a cascade of escapes in nested strings, it will flip the whole f-string to the alternative style.
* **Predictable Nesting:** Eliminates "quote soup" where different nesting levels alternate styles, making the structure easier to parse visually.

## Comparison of Options

To illustrate the differences, here is how each option would handle a complex f-string when **`quote-style = "double"`**:

**Input:** `f'an f-string with "double" quotes and a {'nested string'}'`

| Option | Formatted Result | Logic |
| :--- | :--- | :--- |
| **Option A** (Local Preference) | `f'an f-string with "double" quotes and a {"nested string"}'` | **Outer:** Single (to avoid escapes). **Inner:** Double (per preference). Result: Inconsistent. |
| **Option B** (Strict Preference) | `f"an f-string with \"double\" quotes and a {"nested string"}"` | **Outer:** Double (forced). **Inner:** Double (forced). Result: Consistent but adds escapes (`\"`). |
| **Option C** (Holistic Consistent) | `f'an f-string with "double" quotes and a {'nested string'}'` | **Must be consistent.** Compares "All Double" (2 escapes) vs "All Single" (0 escapes). Result: Consistent and clean. |

---

So far, your poll options A and B have already received some votes, but I think Option C is a compromise between the two, and some people would likely prefer it as well. I hope you can add it to the poll. @matthewlloyd

**Comment by XCanG** [comment]
My personal usecase:

> Today, Ruff's formatter optimizes quotes for **escape minimization**, foregoing the user's preferred quote style for the sake of minimizing escapes:

This is not desired variant as I rather use triple quites `"""` if I want to reduce number of \\, so this would result at maximum with only 1 slash for very last quote as 4 quotes (`"""String with "quotes\""""` exception) otherwise would break syntax.

So option A which will change quotes is not desired output.

**Option B** is best practice here. But I would want to say that may be for an older Python version you may want to keep older behavior until you deprecate support for older python versions. This is also the reason why I was liked the `ruff` as you could specify **target python version** and it was expected that when support for nested formatting strings was introduced, you would switch to that approach as well.

From @vivodi 's Option C is better than Option A, but I still don't like it. If I specify quote for a certain type I would rather expect to use such a quote. I think many people forget about triple quotes that they could be used not only in multiline strings, but single line strings as well. But consistent aproach for either double or single quote is a good approach overall, it just in this case quote style should be auto.
_**So, TL;DR:**_
If specify `quote-style = "double"` then always use double, for nested you could introduce new option and you could add both variants as to specify quote style, e.g. `double` or from Option C add `consistent` variant for nested option, and if `quote-style` is not specified or auto then you could swap quotes and still use option for the nested quotes where you could use this `consistent` variant as well.

**Comment by matthewlloyd** [comment]
> ## Option C: holistic consistent quotes
> `nested-string-quote-style = "consistent"`
> 
> Ensure that nested strings **always use the same quote style as the outermost string**. To keep visual noise to a minimum, the formatter evaluates the entire f-string tree (the outer string plus all nested expressions) and chooses the single quote style that results in the fewest total escapes across the whole structure:

@vivodi Thanks for the comprehensive description of your suggestion.

I would note that Micha has already pushed back on this particular direction in his PR feedback (see https://github.com/astral-sh/ruff/pull/23098#issuecomment-3999666225). He argued we'd have to change the quote selection logic to pick the quote style that results in the fewest total escapes for the entire f/t-string, including all nested strings. That adds quite a bit of complexity for what is essentially a stylistic preference. I agree with him on that.

Furthermore, there is an additional wrinkle that makes this nontrivial.

To my knowledge, the current escape-minimization algorithm optimizes only over the top-level string ('shallow' optimization); nested strings get alternating quotes, so we would make the default value of `nested-string-quote-style` `"alternating"` to represent and preserve the existing behavior.

In order to match the new deep optimization approach you're suggesting for `"consistent"`, where we perform escape-minimization over the entire f/t-string tree, the default quote alternation algorithm would likely also need to be modified to do a similar deep optimization. Otherwise the behavior would be confusingly inconsistent between the two options, with the default `"alternating"` option performing a shallow optimization, and the `"consistent"` option performing a deep optimization.

We would be forced to either (a) allow the behaviors to diverge, which will confuse users and will be tricky to implement, maintain, and document, or (b) unify the behaviors, so that both options perform a deep optimization, which would require a change to the existing default formatting. Neither choice seems desirable.

More importantly, even if we had buy-in from the maintainers for this direction, we don't have a volunteer to implement it - and it may be nontrivial. Given that, it seems unlikely to be pursued.

**Comment by XCanG** [comment]
@matthewlloyd I only have a question of why you are not considering of using `"""` and `'''` as it simplify logic a lot. With triple quotes you only need to check for the very last character to not be the same quote, otherwise escape it. Like if you detect quotes and want to remove \\ then you may put string of the same character (not switching) into triple quotes, e. g. `"some \"test\" string"` -> `"""some "test" string"""`

Do people really don't like triple quotes?

**Comment by matthewlloyd** [comment]
> [@matthewlloyd](https://github.com/matthewlloyd) I only have a question of why you are not considering of using `"""` and `'''` as it simplify logic a lot. 

@XCanG I think this is drifting a bit away from what this issue is trying to solve.

The scope here isn’t about introducing triple quotes as an alternative style, or switching to them when strings contain escaped quotes. Ruff already has a clear model around single vs double quotes via `quote-style`, and this issue is specifically about how that preference should apply inside f-strings (and nested strings).

Adding triple quotes to regular quoted strings would be a much broader formatting change:

* it introduces a third quoting style for normal strings,
* it changes semantics around things like newlines and formatting expectations,
* and it's not something Ruff currently considers for ordinary string normalization.

So while "use triple quotes when needed" is an interesting idea, that would be a separate feature request and off topic for this discussion.

To avoid the poll being buried in the thread, here is a direct link to it: https://github.com/astral-sh/ruff/issues/14118#issuecomment-4069550863

**Comment by MichaReiser** [comment]
I think we should just go with Option A:

* It solves what this issue is about
* I'm not convinced that we should introduce an option that disables minimizing the number of escapes. It has never been requested before. 
* Option C seems an extension of Option A. Something we can iterate on later.

**Comment by TheSkyentist** [comment]
Is Option B still on the table? Given the equal number of votes between the two.

**Comment by MichaReiser** [comment]
I think Option B is out of scope. It proposes how Ruff picks quotes for all strings. This is a much larger change that I consider unlikely to happen anytime soon, because it's not something anyone asked for before it was raised as a potential option.

**Comment by matthewlloyd** [comment]
OK, let's get this long-standing feature request implemented and go with Option A. Users seems divided about what they want but with only 6 votes each there is no statistical significance in either direction. I will update the PR to reflect the new name we chose and address the remaining review comments.

**Comment by MichaReiser** [comment]
Having said that. Feel free to open an issue for Option B and we can consider it long-term.

**Comment by vivodi** [comment]
> Having said that. Feel free to open an issue for Option B and we can consider it long-term.

@matthewlloyd It looks like Option B has more votes now. Would you like to open an issue for it?
